### PR TITLE
feat(face-landmarks) merge face expressions and face centering

### PR DESCRIPTION
**Pull request by quitrk**

### CODE REVIEW

File: react/features/facial-recognition/actions.js [hunk 1]
```diff
@@ -171,73 +208,30 @@ export function startFacialRecognition() {
...
-                        )
+                addFacialExpression(
+                    lastFacialExpression,
+                    duplicateConsecutiveExpressions + 1,
+                    lastFacialExpressionTimestamp
+                )
             );
         }
+
+        clearInterval(webhookSendInterval);
```
Comment (saghul) [inline react/features/facial-recognition/actions.js:222]: I think we should use a single interval and perform multiple detections in a single step.

File: react/features/facial-recognition/actions.js [hunk 2]
```diff
@@ -52,80 +59,97 @@ let duplicateConsecutiveExpressions = 0;
...
 
-        const baseUrl = `${getBaseUrl()}/libs/`;
+        const baseUrl = `${getBaseUrl()}libs/`;
         let workerUrl = `${baseUrl}facial-expressions-worker.min.js`;
 
         const workerBlob = new Blob([ `importScripts("${workerUrl}");` ], { type: 'application/javascript' });
 
         workerUrl = window.URL.createObjectURL(workerBlob);
-        worker = new Worker(workerUrl, { name: 'Facial Expression Worker' });
+        worker = new Worker(workerUrl, { name: 'Face Recognition Worker' });
```
Comment (saghul) [inline react/features/facial-recognition/actions.js:88]: Let's rename this whole feature please. We chose a poor name which is alarming for some, since we are not recognizing faces. We are detecting landmarks. The feature can be "face-landmarks" and the worker friendly name "Face Landmark Detection Worker".

File: react/features/facial-recognition/actions.js [hunk 3]
```diff
@@ -52,80 +59,97 @@ let duplicateConsecutiveExpressions = 0;
...
-            }
-
-            // receives a message with the predicted facial expression.
-            if (type === FACIAL_EXPRESSION_MESSAGE) {
-                sendDataToWorker(worker, imageCapture);
-                if (!value) {
-                    return;
-                }
+            switch (type) {
+            case FACIAL_EXPRESSION_MESSAGE: {
```
Comment (saghul) [inline react/features/facial-recognition/actions.js:93]: Thinking about future extensibility, I think it would be a good idea to use a single type of message with an object as the payload. Each of the landmarks (emotion / experssion and face box currently) would be a key so we process them if we find them.

File: react/features/facial-recognition/actions.js [hunk 4]
```diff
@@ -135,33 +159,46 @@ export function startFacialRecognition() {
         const { recognitionActive } = state['features/facial-recognition'];
 
         if (recognitionActive) {
+            logger.log('Face recognition already active.');
+
             return;
         }
-        const localVideoTrack = getLocalVideoTrack(state['features/base/tracks']);
+
+        const localVideoTrack = track || getLocalVideoTrack(state['features/base/tracks']);
```
Comment (saghul) [inline react/features/facial-recognition/actions.js:167]: On firefox we don't remove tracks when muting, we just mark them as disabled. I wonder if we need an explicit check about it around here.

File: react/features/facial-recognition/facialExpressionsWorker.js [hunk 5]
```diff
@@ -1,120 +1,164 @@
...
+    if (faceDetectionTypes.includes(DETECTION_TYPES.FACE_EXPRESSIONS)) {
+        await detectFaceExpressions({ imageTensor });
+    }
+
+    faceapi.tf.engine().endScope();
+    detectionInProgress = false;
+};
+
+const init = async ({ baseUrl, detectionTypes }) => {
+    initInProgress = true;
```
Comment (saghul) [inline react/features/facial-recognition/facialExpressionsWorker.js:108]: Ditto.

File: react/features/facial-recognition/facialExpressionsWorker.js [hunk 6]
```diff
@@ -1,120 +1,164 @@
...
-                    type: INTERVAL_MESSAGE,
-                    value: timeoutDuration
-                });
-                backendSet = true;
+    }
+
+    // load face detection model
+    if (!modelsLoaded) {
+        try {
+            await faceapi.loadTinyFaceDetectorModel(baseUrl);
```
Comment (saghul) [inline react/features/facial-recognition/facialExpressionsWorker.js:131]: Do we need this even if centering is disabled?

File: react/features/facial-recognition/facialExpressionsWorker.js [hunk 7]
```diff
@@ -1,120 +1,164 @@
...
+let lastValidFaceBox;
+
+const detectFaceBox = async ({ imageTensor, threshold }) => {
+    const detections = await faceapi.detectAllFaces(imageTensor, new faceapi.TinyFaceDetectorOptions());
+
+    if (!detections.length) {
+        return;
+    }
+
+    const faceBox = {
```
Comment (saghul) [inline react/features/facial-recognition/facialExpressionsWorker.js:50]: IF multiple faces are detected we are taking a box that covers them all right?

File: react/features/facial-recognition/functions.js [hunk 8]
```diff
@@ -131,8 +157,69 @@ export async function sendDataToWorker(
         return;
     }
 
+    if (typeof OffscreenCanvas === 'undefined') {
+        const canvas = document.createElement('canvas');
```
Comment (saghul) [inline react/features/facial-recognition/functions.js:161]: I realize I missed this during review last time. I think we should generate a single canvas and reuse it. This way we are creating a ton of DOM nodes which may end up causing problems due to GC.

File: react/features/facial-recognition/middleware.js [hunk 9]
```diff
@@ -5,36 +5,53 @@ import {
...
-    const { enableFacialRecognition } = getState()['features/base/config'];
+    const { enableFacialRecognition, faceCoordinatesSharing } = getState()['features/base/config'];
+    const isEnabled = enableFacialRecognition || faceCoordinatesSharing?.enabled;
 
-    if (!enableFacialRecognition) {
-        return next(action);
-    }
     if (action.type === CONFERENCE_JOINED) {
-        dispatch(loadWorker());
+        if (isEnabled) {
```
Comment (saghul) [inline react/features/facial-recognition/middleware.js:29]: While looking into this I realized I haven't seen how the worker is stopped. Please test this with breakout rooms, since in that scenario we switch from one conference to another and may need to start / tear down the worker.

**Review by saghul — CHANGES_REQUESTED** [review]

File: react/features/facial-recognition/actions.js (hunk 2 above)
Comment (quitrk) [reply react/features/facial-recognition/actions.js:88]: will do so in a future pr, since it will touch other repos, and it will make this one very hard to follow / review

File: react/features/facial-recognition/facialExpressionsWorker.js (hunk 7 above)
Comment (quitrk) [reply react/features/facial-recognition/facialExpressionsWorker.js:48]: yes

File: react/features/facial-recognition/facialExpressionsWorker.js (hunk 6 above)
Comment (quitrk) [reply react/features/facial-recognition/facialExpressionsWorker.js:134]: yup

File: react/features/facial-recognition/facialExpressionsWorker.js (hunk 5 above)
Comment (quitrk) [reply react/features/facial-recognition/facialExpressionsWorker.js:111]: init is only called once, if it's called twice, we have a bigger problem

File: react/features/facial-recognition/facialExpressionsWorker.js [hunk 10]
```diff
@@ -1,120 +1,164 @@
...
 
-    case SET_TIMEOUT : {
-        if (!message.data.imageBitmap || !modelsURL) {
-            self.postMessage({
-                type: FACIAL_EXPRESSION_MESSAGE,
-                value: null
-            });
-        }
+const detect = async ({ image, threshold }) => {
+    detectionInProgress = true;
```
[the remark this answers is not in the document]
Comment (quitrk) [reply react/features/facial-recognition/facialExpressionsWorker.js:82]: I wouldn't be too worried about that

File: react/features/facial-recognition/actions.js (hunk 1 above)
Comment (quitrk) [reply react/features/facial-recognition/actions.js:222]: not sure what you mean, one interval is used for sending frames to worker, the other one is for sending collected expressions on a webhook

File: react/features/facial-recognition/facialExpressionsWorker.js (hunk 5 above)
Comment (saghul) [reply react/features/facial-recognition/facialExpressionsWorker.js:111]: Then maybe we can drop this to make it a bit less confusing for readers? As it stands it looks like it can be called multiple times, and we guard against it.

File: react/features/facial-recognition/actions.js (hunk 3 above)
Comment (quitrk) [reply react/features/facial-recognition/actions.js:93]: arguably, processing these messages take different code paths, and you'd only replace the switch with if statements that check for the respective keys in order to take the appropriate execution path.

File: react/features/facial-recognition/facialExpressionsWorker.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (saghul) [reply react/features/facial-recognition/facialExpressionsWorker.js:82]: Yeah it's no biggie, just wondering if Bad Things (TM) can happen if we have 2 concurrent detects running.

File: react/features/facial-recognition/actions.js (hunk 2 above)
Comment (saghul) [reply react/features/facial-recognition/actions.js:88]: Git is quite smart if you rename files, but ok. What other repos do we need to modify?

File: react/features/facial-recognition/actions.js (hunk 4 above)
Comment (quitrk) [reply react/features/facial-recognition/actions.js:167]: but this function only runs when we start (e.g. unmute camera).

File: react/features/facial-recognition/actions.js (hunk 1 above)
Comment (saghul) [reply react/features/facial-recognition/actions.js:222]: Sorry, I realize I left the comment in the wrong place. I see you've made all landmark detections work on a single timer. Is that something desirable long term? Do we need to detect emotions with the same frequency as centering?

File: react/features/facial-recognition/middleware.js (hunk 9 above)
Comment (quitrk) [reply react/features/facial-recognition/middleware.js:29]: good catch. tested across breakout rooms, we can keep reusing same worker instance, added a check in order to prevent reinitializing it.

File: react/features/facial-recognition/actions.js (hunk 1 above)
Comment (quitrk) [reply react/features/facial-recognition/actions.js:228]: ideally we would do both detections for a given frame, so we can either use the min or the max interval from the two. Currently the minimum is set.

File: react/features/facial-recognition/facialExpressionsWorker.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (quitrk) [reply react/features/facial-recognition/facialExpressionsWorker.js:82]: not really, from face expressions standpoint the result would be the same, as for face centering, the slowest processing would end up overriding the coordinates.

File: react/features/facial-recognition/actions.js (hunk 2 above)
Comment (quitrk) [reply react/features/facial-recognition/actions.js:94]: problem is we also need to modify variable names, functions, comments, etc

File: react/features/facial-recognition/facialExpressionsWorker.js (hunk 5 above)
Comment (quitrk) [reply react/features/facial-recognition/facialExpressionsWorker.js:111]: actually I can make use of other flags for that, ill remove this

File: react/features/facial-recognition/actions.js (hunk 1 above)
Comment (saghul) [reply react/features/facial-recognition/actions.js:228]: Maybe we can revisit in the future, if any of the landmarks poses a problem. WDYT @gabiborlea ?

**Comment by saghul** [comment]
@gabiborlea can you pl review too?

**Review by gabiborlea — CHANGES_REQUESTED** [review]

File: webpack.config.js [hunk 11]
```diff
@@ -384,16 +384,6 @@ module.exports = (_env, argv) => {
...
-            },
-            plugins: [
-                ...config.plugins,
-                ...getBundleAnalyzerPlugin(analyzeBundle, 'face-centering-worker')
-            ],
-            performance: getPerformanceHints(perfHintOptions, 500 * 1024)
-        }),
         Object.assign({}, config, {
             entry: {
                 'facial-expressions-worker': './react/features/facial-recognition/facialExpressionsWorker.js'
```
[the remark this answers is not in the document]
Comment (quitrk) [reply webpack.config.js:389]: the whole renaming part will come in a future PR

File: react/features/facial-recognition/facialExpressionsWorker.js [hunk 12]
```diff
@@ -1,120 +1,158 @@
...
+    if (lastValidFaceBox && Math.abs(lastValidFaceBox.left - faceBox.left) < threshold) {
+        return;
     }
+
+    lastValidFaceBox = faceBox;
+
+    self.postMessage({
+        type: FACE_BOX_MESSAGE,
+        value: faceBox
+    });
```
[the remark this answers is not in the document]
Comment (quitrk) [reply react/features/facial-recognition/facialExpressionsWorker.js:60]: we already talked about this with Saul, the reasoning for sending separate messages was to keep the handling logic cleaner, do you see a performance issue with this approach?

File: react/features/facial-recognition/facialExpressionsWorker.js [hunk 13]
```diff
@@ -1,120 +1,158 @@
...
+    detectionInProgress = true;
+    faceapi.tf.engine().startScope();
+
+    const imageTensor = faceapi.tf.browser.fromPixels(image);
+
+    if (faceDetectionTypes.includes(DETECTION_TYPES.FACE_EXPRESSIONS)) {
+        detections = await faceapi.detectAllFaces(
+            imageTensor,
+            new faceapi.TinyFaceDetectorOptions()
         ).withFaceExpressions();
```
[the remark this answers is not in the document]
Comment (quitrk) [reply react/features/facial-recognition/facialExpressionsWorker.js:86]: we already reuse that if it exists: `detections ? detections.map(d => d.detection)...`

File: react/features/facial-recognition/facialExpressionsWorker.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (gabiborlea) [reply react/features/facial-recognition/facialExpressionsWorker.js:60]: Not sure how costly is to send a message to the main thread, but I guess the performance would not be affected.

[the remark this answers is not in the document]
Comment (quitrk) [reply react/features/facial-recognition/facialExpressionsWorker.js]: updated

**Comment by gabiborlea** [comment]
@saghul seems to be fine

File: react/features/facial-recognition/actions.js [hunk 14]
```diff
@@ -52,80 +57,99 @@ let duplicateConsecutiveExpressions = 0;
...
+
+                dispatch({
+                    type: UPDATE_FACE_COORDINATES,
+                    faceBox,
+                    id: localParticipant.id
+                });
+            }
         };
+
+        const { enableFacialRecognition, faceCoordinatesSharing } = getState()['features/base/config'];
```
Comment (saghul) [inline react/features/facial-recognition/actions.js:130]: Note for the next phase: group these in some `faceLandmarks` object.

File: react/features/facial-recognition/actions.js [hunk 15]
```diff
@@ -52,80 +57,99 @@ let duplicateConsecutiveExpressions = 0;
...
+        const detectionTypes = [
+            faceCoordinatesSharing?.enabled && DETECTION_TYPES.FACE_BOX,
+            enableFacialRecognition && DETECTION_TYPES.FACE_EXPRESSIONS
+        ].filter(Boolean);
```
Comment (saghul) [inline react/features/facial-recognition/actions.js:134]: Do we still want to initialize the worker if both are disabled and the array is empty?

File: react/features/facial-recognition/actions.js [hunk 16]
```diff
@@ -52,80 +57,99 @@ let duplicateConsecutiveExpressions = 0;
...
+
         dispatch(startFacialRecognition());
     };
 }
 
 /**
  * Starts the recognition and detection of face expressions.
  *
- * @param  {Object} stream - Video stream.
+ * @param {Track | undefined} track - Track for which to start detecting faces.
```
Comment (saghul) [inline react/features/facial-recognition/actions.js:149]: When would the track de defined and when not?

**Review by saghul — APPROVED** [review]
LGTM with a small question.

File: react/features/facial-recognition/actions.js (hunk 15 above)
Comment (quitrk) [reply react/features/facial-recognition/actions.js:134]: the check for initializing the worker is done in the middleware, this is just for passing the array, at this point we know that at least one of them is active

File: react/features/facial-recognition/actions.js (hunk 14 above)
Comment (quitrk) [reply react/features/facial-recognition/actions.js:130]: yup

File: react/features/facial-recognition/actions.js (hunk 16 above)
Comment (quitrk) [reply react/features/facial-recognition/actions.js:149]: we only pass the track when we intercept it changes in the middleware, at which point it wouldn't be available yet in the state

**Comment by quitrk** [comment]
Jenkins please test this please
